### DISCUSSION TITLE
Harmonic Davidson approach

### DISCUSSION THREAD

**Discussion by wltr-tlrc** [block-hczhai/block2-preview#96] [Q&A] [answered]
Hello dear developers, 
and thank you for building up this great package!

In the descriptive article of the code [10.1063/5.0180424](https://doi.org/10.1063/5.0180424), the authors mentioned the possibility to use the Harmonic Davidson approach to target specific high-energy eigenstate without computing all the low lying states as opposite of the other implemented methods like state-averaged, pojected orthogonalization, ecc...   
Unfortunately, I wasn't able to find anywhere in the documentation and in the code-base an example on how one could use this specific feature.

In particular, how I should change the example below in which instead of calculating the first 10 excited states of the Heisenberg model with a state-averaged approach, I could target only the 10th excited state with the Harmonic Davidson method ?! 

```
import numpy as np
from pyblock2.driver.core import DMRGDriver, SymmetryTypes
from pyblock2._pyscf.ao2mo import integrals as itg
import psutil
num_cpus = psutil.cpu_count(logical=False)

L = 22 
heis_twos = 1

driver = DMRGDriver(scratch="./tmp", symm_type=SymmetryTypes.SGB, n_threads=num_cpus)
driver.initialize_system(n_sites=L, heis_twos=heis_twos, heis_twosz=0)

b = driver.expr_builder()
for i in range(L - 1):
    b.add_term("PM", [i, i + 1], 0.5)
    b.add_term("MP", [i, i + 1], 0.5)
    b.add_term("ZZ", [i, i + 1], 1.0)

heis_mpo = driver.get_mpo(b.finalize(), iprint=0)

ket = driver.get_random_mps(tag="KET", bond_dim=250, nroots=10)
bond_dims = [250] * 4 + [500] * 4
noises = [1e-5] * 4 + [1e-6] * 2 + [0]
thrds = [1e-6] * 4 + [1e-8] * 4

energies = driver.dmrg(
    heis_mpo,
    ket,
    n_sweeps=8,
    bond_dims=bond_dims,
    noises=noises,
    thrds=thrds,
    cutoff=1E-24,
    iprint=1,
)
energies = [x / L for x in energies]
print('State-averaged MPS energies = [%s]' % " ".join("%20.15f" % x for x in energies))

kets = [driver.split_mps(ket, ir, tag="KET-%d" % ir) for ir in range(ket.nroots)]
for ir in range(ket.nroots):
    energy = driver.dmrg(heis_mpo, kets[ir], n_sweeps=1, bond_dims=bond_dims, noises=noises,
        thrds=thrds, iprint=0, proj_weights=[5.0] * ir, proj_mpss=kets[:ir])
    print('State-specific MPS E[%d] = %20.15f' % (ir, energy/L))
```

**Comment by hczhai** [comment, marked as answer]
Hi, thanks for your interest in using block2.

With Harmonic Davidson approach, you can target excited state which is close to a given energy (via the ``davidson_shift`` argument of ``driver.dmrg``) instead of the $n$ th state. The following lines will allow you to do that:

```python3
ket = driver.get_random_mps(tag="KET", bond_dim=250, nroots=1)
bond_dims = [250] * 4 + [500] * 4
noises = [1e-5] * 4 + [1e-6] * 2 + [0]
thrds = [1e-12] * 8

energy = driver.dmrg(
    heis_mpo, ket, n_sweeps=8, bond_dims=bond_dims,
    noises=noises, thrds=thrds, cutoff=1E-24,
    dav_type="Harmonic|CloseTo", davidson_shift=-8.88, iprint=2,
)
print('Energy = [%s]' % " ".join("%20.15f" % x for x in [energy]))
```

Note that it is recommended to use a tighter Davidson threshold (1E-12) for this case. If you compare this with your original output, you can see that this gives you the 7th excited state (E=-8.88986002). As stated in the paper, the current implementation of Harmonic Davidson is not stable. You may just do a normal Davidson and directly target an excited state, which seems to be more stable for your case:

```python3
ket = driver.get_random_mps(tag="KET", bond_dim=250, nroots=1)
bond_dims = [250] * 4 + [500] * 4
noises = [1e-5] * 4 + [1e-6] * 2 + [0]
thrds = [1e-12] * 8

energy = driver.dmrg(
    heis_mpo, ket, n_sweeps=8, bond_dims=bond_dims,
    noises=noises, thrds=thrds, cutoff=1E-24,
    dav_type="CloseTo", davidson_shift=-8.88, iprint=2,
)
print('Energy = [%s]' % " ".join("%20.15f" % x for x in [energy]))
```

**Reply by hczhai** [reply]
An up-to-date version of ``block2`` is recommended for the above task: ``python3 -m pip install block2==0.5.3rc13 --extra-index-url=https://block-hczhai.github.io/block2-preview/pypi/``

**Comment by wltr-tlrc** [comment]
Hi @hczhai, thanks a lot for your explanation. 
I've missed that argument because I was using the version `0.5.2` of the package and for that the `DMRGDriver.dmrg()`  didn't yet have that specific keyword argument.
